File: java/ProductofArrayExceptSelf.java
/*
 * Filename: ProductofArrayExceptSelf.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 15 Jul 2015 09:38:39 PM EDT
 */

public class Solution {
	public int[] productExceptSelf(int[] nums) {
		if (nums == null || nums.length == 0) {
			return new int[0];
		}

		int n = nums.length;
		int[] res = new int[n];
		
		for (int i = 0; i < n; i++) {
			if (i == 0) {
				res[i] = 1;
			} else {
				res[i] = res[i-1] * nums[i-1];
			}
		}

		int prod = 1;
		for (int i = n-1; i >= 0; i--) {
			if (i != n-1) {
				prod *= nums[i+1];
			}

			res[i] *= prod;
		}
		return res;
	}
}

File: cpp/product_of_array_except_self.cc
/*
 * Filename: product_of_array_except_self.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 14 Aug 2015 12:30:54 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		vector<int> productExceptSelf(vector<int>& nums) {
			if (nums.empty())
			{
				vector<int> res;
				return res;
			}

			int n = nums.size();
			vector<int> res (n);
			vector<int> forward (n, 1);
			vector<int> backward (n, 1);

			for (int i = 1; i < n; i++)
			{
				forward[i] = forward[i-1] * nums[i-1];
			}

			for (int i = n-2; i >= 0; i--)
			{
				backward[i] = backward[i+1] * nums[i+1];
			}

			for (int i = 0; i < n; i++)
			{
				res[i] = forward[i] * backward[i];
			}

			return res;
		}
};
